manifests/composer.pp

```diff
@@ -38,9 +38,31 @@
 ) inherits php::params {
   assert_private()
 
+  case $channel {
```

Comment:
Usage of a selector would be more concise:

```puppet
$main_source = $channel ? {
  stable  => '...',
  preview => '...',
  # ...
}
```

Reply:
my thoughts would be:

the `$channel` variable is a custom type (basically Enum), so we could skip the error handling, remove the case statement and instead use a hash.

i would add a new variable called `$channel_sources` and put the predefined channel sources in  as hash. so channel sources can easily be overwritten if required.

https://github.com/voxpupuli/puppet-php/blob/master/data/default.yaml
```yaml
channel_sources:
  stable:   'https://getcomposer.org/download/latest-stable/composer.phar'
  preview:  'https://getcomposer.org/download/latest-preview/composer.phar'
  snapshot: 'https://getcomposer.org/composer.phar'
  1:        'https://getcomposer.org/download/latest-1.x/composer.phar'
  2:        'https://getcomposer.org/download/latest-2.x/composer.phar'
```

Reply:
Are some URL OS-dependant?  If not, I am not sure the ability to override them in hiera is needed: the `$source` parameter already allows to set a custom value and the module alway provide the well-known values regardless of what people put in hiera.

Reply:
to override something can always be helpful. for example this or another module (can't remember) had the apt repo key hardcoded like this. this is not a problem on high active modules where PR are merged fast but on repos where fixes are only merged very slow it requires you to fork the repo and fix the key yourself. to allow the user to override static values is, in my perspective, a best practice.

a usecase could be that i have an internal mirror of all composer channels, with static config in hiera i simply can choose my own channelurls without touch the modules code.

Reply:
I have made the modification suggested by @c33s. I have made some tests on a real puppet installation and the result is correct.
If it's ok, I can made a rebase to merge all my commit into one.